tools/autograd/gen_variable_type.py

```diff
@@ -185,49 +198,117 @@ def should_trace(declaration):
     return True
 
 
-def record_trace_outputs(declaration):
-    if declaration['name'].endswith('_out'):
-        output_names = [arg['name'] for arg in declaration['arguments'] if arg.get('output', False)]
-    else:
-        output_names = [r['name'] for r in declaration['returns']]
-    return ['jit::tracer::addOutput(node, {});'.format(n) for n in output_names]
+def is_out_overload(declaration):
+    return declaration['api_name'].endswith('_out')
 
 
-def format_trace(declaration):
-    local = {}
-    local['trace_name'] = trace_name = uninplace_api_name(declaration['api_name'])
+def format_postrecord_trace(declaration):
+    # For outplacing ops, *_out overloads require special handling to move the
+    # output *argument* to a return value
+    if is_out_overload(declaration):
+        output_names_outplace = [arg['name'] for arg in declaration['arguments'] if arg.get('output', False)]
+        output_names_inplace = [r['name'] for r in declaration['returns']]
+
+        if output_names_outplace == output_names_inplace:
+            outputs = ['jit::tracer::addOutput(node, {});'.format(n) for n in output_names_outplace]
+            return POST_RECORD_TRACE.substitute(add_trace_outputs=outputs)
+
+        local = {}
+        local['cond'] = 'force_outplace'
+        local['true'] = ['jit::tracer::addOutput(node, {});'.format(n) for n in output_names_outplace]
+        local['false'] = ['jit::tracer::addOutput(node, {});'.format(n) for n in output_names_inplace]
+        selection = SELECT.substitute(local)
+        return POST_RECORD_TRACE.substitute(add_trace_outputs=selection)
+
+    output_names = [r['name'] for r in declaration['returns']]
+    outputs = ['jit::tracer::addOutput(node, {});'.format(n) for n in output_names]
+    return POST_RECORD_TRACE.substitute(add_trace_outputs=outputs)
+
+
+def format_trace_op_name(declaration):
+    def rename(trace_name):
+        if trace_name in RENAME_TRACE:
+            trace_name = RENAME_TRACE[trace_name]
+        return trace_name
```

Comment:
This is just `RENAME_TRACE.get(trace_name, trace_name)`?

Reply:
python!